### PR TITLE
bpo-2504: Add pgettext() and variants to gettext

### PR DESCRIPTION
**Pull request by csabella**
Original patch by Franz Glasner.

https://bugs.python.org/issue2504

### CODE REVIEW

File: Doc/library/gettext.rst [hunk 1]
```diff
@@ -379,8 +418,37 @@ unexpected, or if other problems occur while reading the file, instantiating a
              n) % {'num': n}
 
 
+   .. method:: pgettext(context, message)
+
+      Look up the *context* and *message* id in the catalog and return the
+      corresponding message string, as an 8-bit string encoded with the
```
Comment (serhiy-storchaka) [inline Doc/library/gettext.rst]: This is not true in Python 3. The result is a Unicode string, not an encoded 8-bit string. Only `l*gettext` functions return bytes, and this makes them virtually useless.

**Comment by A3880036** [comment]
Accept

**Comment by damufo** [comment]
Hello:
How is this?
I think it would be interesting to complete it.
Being able to use context in gettext would allow for example to specify the gender.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Remove `l*gettext()` functions and fix the documentation.

**Comment by csabella** [comment]
I have made the requested changes; please review again.

File: Doc/library/gettext.rst [hunk 2]
```diff
@@ -316,7 +344,7 @@ are the methods of :class:`!NullTranslations`:
       If the *names* parameter is given, it must be a sequence containing the
       names of functions you want to install in the builtins namespace in
       addition to :func:`_`.  Supported names are ``'gettext'``, ``'ngettext'``,
-      ``'lgettext'`` and ``'lngettext'``.
+      ``'pgettext'``, ``'lgettext'``, and ``'lngettext'``.
```
Comment (serhiy-storchaka) [inline Doc/library/gettext.rst]: Why is 'npgettext' omitted?

Maybe add `versionchanged`?

File: Doc/library/gettext.rst [hunk 3]
```diff
@@ -394,6 +422,32 @@ unexpected, or if other problems occur while reading the file, instantiating a
...
+      :meth:`pgettext` method.  Otherwise, the *message* id is returned.
+
+      .. versionadded:: 3.8
+
+
+   .. method:: npgettext(context, singular, plural, n)
+
+      Do a plural-forms lookup of a message id.  *singular* is used as the
+      message id for purposes of lookup in the catalog, while *n* is used to
+      determine which  plural form to use.  The returned message string is a
```
Comment (serhiy-storchaka) [inline Doc/library/gettext.rst]: I think the sentence about a Unicode string can be removed.

File: Lib/gettext.py [hunk 4]
```diff
@@ -493,6 +507,29 @@ def ngettext(self, msgid1, msgid2, n):
...
+        if tmsg is missing:
+            if self._fallback:
+                return self._fallback.pgettext(context, message)
+            return message
+        return tmsg
+
+    def npgettext(self, context, msgid1, msgid2, n):
+        ctxt_msg_id = self.CONTEXT % (context, msgid1)
+        try:
+            tmsg = self._catalog[(ctxt_msg_id, self.plural(n))]
```
Comment (serhiy-storchaka) [inline Lib/gettext.py]: Parenthesis are not necessary: `self._catalog[ctxt_msg_id, self.plural(n)]`.

File: Tools/i18n/msgfmt.py [hunk 5]
```diff
@@ -32,28 +33,34 @@
...
+
+def add(ctxt, id, str, fuzzy):
     "Add a non-fuzzy translation to the dictionary."
     global MESSAGES
     if not fuzzy and str:
-        MESSAGES[id] = str
+        if ctxt is None:
+            MESSAGES[id] = str
+        else:
+            MESSAGES["%s\x04%s" % (ctxt, id)] = str
```
Comment (serhiy-storchaka) [inline Tools/i18n/msgfmt.py]: Is not `id` a bytes string?

File: Tools/i18n/msgfmt.py [hunk 6]
```diff
@@ -183,7 +198,9 @@ def make(filename, outfile):
         if not l:
             continue
         l = ast.literal_eval(l)
-        if section == ID:
+        if section == CTXT:
+            msgctxt += l
```
Comment (serhiy-storchaka) [inline Tools/i18n/msgfmt.py]: Encode?

**Comment by serhiy-storchaka** [comment]
Please add also a What's New entry.

File: Doc/library/gettext.rst (hunk 3 above)
Comment (csabella) [reply Doc/library/gettext.rst]: Removed.  I think it was originally included as a copy of `ngettext`.

File: Tools/i18n/msgfmt.py (hunk 5 above)
Comment (csabella) [reply Tools/i18n/msgfmt.py]: Thanks!  Fixed.

File: Tools/i18n/msgfmt.py (hunk 6 above)
Comment (csabella) [reply Tools/i18n/msgfmt.py]: Thanks!

File: Doc/library/gettext.rst [hunk 7]
```diff
@@ -359,6 +359,9 @@ are the methods of :class:`!NullTranslations`:
       This puts :func:`_` only in the module's global namespace and so only
       affects calls within this module.
 
+      .. versionchanged:: 3.8
+      Added ``'pgettext'`` and ``'npgettext'``.
```
Comment (serhiy-storchaka) [inline Doc/library/gettext.rst]: Bad indentation.

File: Doc/whatsnew/3.8.rst [hunk 8]
```diff
@@ -128,6 +128,13 @@ asyncio
 On Windows, the default event loop is now :class:`~asyncio.ProactorEventLoop`.
 
 
+gettext
+-------
+
+Added :func:`~gettext.pgettext` and its variants.
+(Contributed by Franz Glasner in :issue:`2504`.)
```
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.8.rst]: I think it is worth to add Éric and you as co-authors.

File: Doc/library/gettext.rst (hunk 7 above)
Comment (csabella) [reply Doc/library/gettext.rst]: Sorry about that.

File: Doc/whatsnew/3.8.rst (hunk 8 above)
Comment (csabella) [reply Doc/whatsnew/3.8.rst]: OK.

**Review by merwok — APPROVED** [review]

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**